### PR TITLE
Make package and toolfile for jemalloc with heap profiling enabled.

### PR DESCRIPTION
**Pull request by gartung**
The toolfile is added to available since two jemalloc tool config can't be enable at once.

### CODE REVIEW

**Comment by gartung** [comment]
please test

File: scram-tools.file/tools/jemalloc-prof/jemalloc-prof.xml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+<tool name="jemalloc-debug" version="@TOOL_VERSION@">
```
Comment (smuzaffar) [inline scram-tools.file/tools/jemalloc-prof/jemalloc-prof.xml]: @gartung , please fix the tool file. The name should be `jemalloc-prof` and we should have `<lib name="jemalloc-prof"/>`

File: scram-tools.file/tools/jemalloc-prof/jemalloc-prof.xml [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
+  <lib name="jemalloc"/>
+  <client>
+    <environment name="JEMALLOC_BASE" default="@TOOL_ROOT@"/>
```
Comment (smuzaffar) [inline scram-tools.file/tools/jemalloc-prof/jemalloc-prof.xml]: replace `JEMALLOC_BASE` with `JEMALLOC_PROF_BASE` every where in the tool file

**Comment by gartung** [comment]
please test

**Comment by gartung** [comment]
please test

**Comment by gartung** [comment]
```
MALLOC_CONF=prof_leak:true,lg_prof_sample:20,prof_final:true  cmsRun step3.py
<jemalloc>: Leak approximation summary: ~320852416 bytes, ~731 objects, >= 239 contexts
<jemalloc>: Run jeprof on "jeprof.11286.0.f.heap" for leak detail
```

**Comment by gartung** [comment]
```
/build/gartung/cmsbuild/el8_amd64_gcc10/external/jemalloc-prof/5.2.1-b2e0b7da8e8c66672d3f91727db4b879/bin/jeprof --show_bytes `which cmsRun` jeprof.11286.0.f.heap 
Using local file /build/gartung/CMSSW_12_5_0_pre1/bin/el8_amd64_gcc10/cmsRun.
Using local file jeprof.11286.0.f.heap.
Welcome to jeprof!  For help, type 'help'.
(jeprof) top
Total: 338092591 B
215677567  63.8%  63.8% 215677567  63.8% je_prof_backtrace
32789768   9.7%  73.5% 36460759  10.8% clang::ASTReader::ReadASTBlock
14681432   4.3%  77.8% 14681432   4.3% TStorage::ObjectAlloc
11537152   3.4%  81.2% 12585760   3.7% std::filesystem::__cxx11::path::_List::_Impl::copy (inline)
 8395266   2.5%  83.7%  8395266   2.5% llvm::SmallDenseMap::grow
 5243680   1.6%  85.3% 15734177   4.7% clang::DeclContext::makeDeclVisibleInContextImpl
 4194536   1.2%  86.5%  4194536   1.2% google::protobuf::Arena::CreateMaybeMessage
 4194528   1.2%  87.8%  4194528   1.2% RPCGeometryBuilderFromCondDB::build
 4194464   1.2%  89.0%  4194464   1.2% TList::NewLink
 4194440   1.2%  90.2%  4194440   1.2% std::__cxx11::basic_string::_M_construct [clone .isra.0]
```

**Comment by gartung** [comment]
```
Heap-Profile Options:
   --inuse_space       Display in-use (mega)bytes [default]
   --inuse_objects     Display in-use objects
   --alloc_space       Display allocated (mega)bytes
   --alloc_objects     Display allocated objects
   --show_bytes        Display space in bytes
   --drop_negative     Ignore negative differences
```

**Comment by gartung** [comment]
please test

**Comment by gartung** [comment]
abort

**Comment by gartung** [comment]
please test

File: jemalloc-prof.spec [hunk 3]
```diff
@@ -0,0 +1,25 @@
+%define jemalloc_version 5.2.1
```
Comment (smuzaffar) [inline jemalloc-prof.spec]: @gartung why not use `## INCLUDE jemalloc-common` just like jemalloc-debug https://github.com/cms-sw/cmsdist/blob/IB/CMSSW_12_5_X/master/jemalloc-debug.spec#L1-L2 ?

Comment (gartung) [reply jemalloc-prof.spec]: I needed to add `Requires: libunwind`.

Comment (gartung) [reply jemalloc-prof.spec]: The `Requires:` needs to come before `prep:` as far as I know.

Comment (smuzaffar) [reply jemalloc-prof.spec]: can you do that by adding `Requires: libunwind` before the `## INCLUDE jemalloc-common` line e.g.
```
Requires: libunwind
## INCLUDE jemalloc-common
```
should work

Comment (gartung) [reply jemalloc-prof.spec]: Yes that worked. Now I need to figure out why libunwind is not linked to libjemalloc.

**Comment by gartung** [comment]
please test

**Comment by gartung** [comment]
```
MALLOC_CONF=prof_leak:true,lg_prof_sample:10,prof_final:true  cmsRun step3.py
<jemalloc>: Leak approximation summary: ~280717645 bytes, ~85649 objects, >= 28584 contexts
<jemalloc>: Run jeprof on "jeprof.12593.0.f.heap" for leak detail
```

**Comment by gartung** [comment]
Example of in use objects heap profile sorted by self cost. Columns are self cost, self cost %, cumulative %, self cost, self cost %
```
jeprof --text --inuse_objects  `which cmsRun` jeprof.12593.0.f.heap

Total: 692034 objects
   85823  12.4%  12.4%    85823  12.4% TStorage::ObjectAlloc
   58285   8.4%  20.8%    58285   8.4% google::protobuf::Arena::CreateMaybeMessage
   45570   6.6%  27.4%    45570   6.6% std::__cxx11::basic_string::_M_construct [clone .isra.0]
   43751   6.3%  33.7%    43751   6.3% TList::NewLink
   40205   5.8%  39.5%    45955   6.6% google::protobuf::internal::ArenaStringPtr::Set@3caaba
   35431   5.1%  44.7%    35431   5.1% je_prof_backtrace
   35152   5.1%  49.7%    35152   5.1% std::__cxx11::basic_string::_M_construct
   24282   3.5%  53.2%    24282   3.5% clang::DeclContext::CreateStoredDeclsMap
   16706   2.4%  55.7%    16706   2.4% google::protobuf::internal::ArenaStringPtr::Set@3cad7e
   16532   2.4%  58.1%    45901   6.6% std::filesystem::__cxx11::path::_List::_Impl::copy (inline)
   16404   2.4%  60.4%    16404   2.4% google::protobuf::internal::RepeatedPtrFieldBase::InternalExtend
   15372   2.2%  62.6%    15372   2.2% google::protobuf::RepeatedField::Reserve
   14725   2.1%  64.8%    44270   6.4% clang::DeclContext::makeDeclVisibleInContextImpl
   14315   2.1%  66.8%    14333   2.1% clang::StoredDeclsList::AddSubsequentDecl
   14216   2.1%  68.9%    21518   3.1% RPCGeometryBuilderFromCondDB::build
   12647   1.8%  70.7%    12647   1.8% TString::Init
   12366   1.8%  72.5%    12366   1.8% TString::Clobber
   10384   1.5%  74.0%   135479  19.6% tensorflow::kernel_factory::OpKernelRegistrar::OpKernelRegistrar
   10118   1.5%  75.5%   131385  19.0% tensorflow::kernel_factory::OpKernelRegistrar::InitInternal
    9826   1.4%  76.9%    47452   6.9% TStreamerInfo::Compile
    8450   1.2%  78.1%     8450   1.2% TClassTable::AddAlternate
    7886   1.1%  79.3%     7886   1.1% THashTable::THashTable
    7622   1.1%  80.4%     7644   1.1% clang::serialization::MultiOnDiskHashTable::add
    6704   1.0%  81.3%     6704   1.0% TStreamerInfo::AddWriteAction
    6324   0.9%  82.2%     6768   1.0% TStreamerInfo::AddReadAction
    6220   0.9%  83.1%     6220   0.9% TObjArray::Init
    5912   0.9%  84.0%     5912   0.9% llvm::SmallDenseMap::grow
    5720   0.8%  84.8%     5720   0.8% TString::Replace
    5506   0.8%  85.6%     6059   0.9% RPCRollSpecs::RPCRollSpecs
    5175   0.7%  86.4%     5201   0.8% Xbyak::LabelManager::define_inner
    4485   0.6%  87.0%     4485   0.6% google::protobuf::internal::StringTypeHandler::New[abi:cxx11]
    4386   0.6%  87.6%     4386   0.6% GEMEtaPartitionSpecs::GEMEtaPartitionSpecs
    4170   0.6%  88.2%     4170   0.6% TStreamerInfo::AddWriteTextAction
    3947   0.6%  88.8%     3947   0.6% std::__cxx11::basic_string::_M_mutate
    3843   0.6%  89.4%     3843   0.6% TStreamerInfo::AddWriteMemberWiseVecPtrAction
    3699   0.5%  89.9%     3699   0.5% GEMGeometryBuilderFromCondDB::boundPlane
    3624   0.5%  90.4%     3624   0.5% TStreamerInfoActions::TConfiguredAction GetCollectionReadAction@1c8360
    3605   0.5%  91.0%     3605   0.5% TExMap::TExMap
    3425   0.5%  91.4%     4175   0.6% llvm::BitstreamCursor::ReadAbbrevRecord
    3324   0.5%  91.9%     3324   0.5% std::_Deque_base::_M_initialize_map
    3308   0.5%  92.4%    51494   7.4% tensorflow::OpRegistry::RegisterAlreadyLocked
    3100   0.4%  92.9%     3100   0.4% TStreamerInfo::AddReadTextAction
    3098   0.4%  93.3%     5747   0.8% TClass::AddClass
    2839   0.4%  93.7%     5213   0.8% std::vector::_M_realloc_insert
    2827   0.4%  94.1%     2861   0.4% clang::ASTReader::ReadToken
    2783   0.4%  94.5%    10325   1.5% GEMGeometryBuilderFromCondDB::buildEtaPartition
    2685   0.4%  94.9%     2685   0.4% TStreamerInfoActions::TConfiguredAction GetCollectionReadAction@1c2f00
    2441   0.4%  95.3%    96263  13.9% TClass::Init
    1878   0.3%  95.5%     4818   0.7% TCling::DataMemberInfo_FactoryCopy
    1788   0.3%  95.8%     4297   0.6% TClassTable::Add
    1755   0.3%  96.0%    11653   1.7% tensorflow::XlaOpRegistry::RegisterCompilationKernels
    1551   0.2%  96.3%     2100   0.3% clang::ASTContext::mergeDefinitionIntoModule
    1353   0.2%  96.5%     1353   0.2% TClassTable::FindElementImpl
```

**Comment by gartung** [comment]
Examples of in use space bytes sorted by self cost. 

```
jeprof --text --inuse_space --show_bytes --exclude="(je_prof_backtrace|prof_alloc_prep)" `which cmsRun` jeprof.12593.0.f.heap 

Total: 338132989 B
44716719  13.2%  13.2% 44716719  13.2% clang::Decl::operator new@2b674e0
32843108   9.7%  22.9% 37092336  11.0% clang::ASTReader::ReadASTBlock
23715881   7.0%  30.0% 23715881   7.0% clang::ASTContext::CreateTypeSourceInfo
18694354   5.5%  35.5% 18694354   5.5% llvm::BumpPtrAllocatorImpl::Allocate [clone .constprop.1]
14299996   4.2%  39.7% 14299996   4.2% llvm::FoldingSetBase::FoldingSetBase
12525737   3.7%  43.4% 20922540   6.2% clang::ASTContext::getTemplateSpecializationType@2a71930
11707677   3.5%  46.9% 11707677   3.5% TStorage::ObjectAlloc
11316960   3.3%  50.2% 11327200   3.3% clang::Decl::operator new@2b60170
10920290   3.2%  53.5% 10920290   3.2% llvm::SmallVectorBase::grow_pod
10271584   3.0%  56.5% 10276739   3.0% llvm::BumpPtrAllocatorImpl::Allocate
 9383516   2.8%  59.3%  9383516   2.8% llvm::BumpPtrAllocatorImpl::Allocate [clone .constprop.0]
 8391287   2.5%  61.7%  9555302   2.8% std::filesystem::__cxx11::path::_List::_Impl::copy (inline)
 7170499   2.1%  63.9%  7170499   2.1% llvm::SmallDenseMap::grow
 6032609   1.8%  65.7%  6032609   1.8% google::protobuf::Arena::CreateMaybeMessage
 5242880   1.6%  67.2%  5242880   1.6% clang::ASTContext::getTrivialTypeSourceInfo
 4980736   1.5%  68.7%  5505024   1.6% clang::MakeDeductionFailureInfo
 4712047   1.4%  70.1% 13105864   3.9% clang::DeclContext::makeDeclVisibleInContextImpl
 4580892   1.4%  71.4%  4678440   1.4% clang::StoredDeclsList::AddSubsequentDecl
 3500144   1.0%  72.5%  3500144   1.0% TList::NewLink
 3457524   1.0%  73.5%  3457524   1.0% llvm::FoldingSetBase::GrowBucketCount
 3381954   1.0%  74.5%  3381954   1.0% std::__cxx11::basic_string::_M_construct [clone .isra.0]
 2818048   0.8%  75.3%  2818048   0.8% clang::UnresolvedLookupExpr::Create@2c2c5a0
 2810531   0.8%  76.1%  2810531   0.8% llvm::DenseMap::grow
 2789452   0.8%  77.0%  2789452   0.8% clang::TemplateArgumentList::CreateCopy
 2703360   0.8%  77.8%  2703360   0.8% clang::Stmt::operator new
 2408609   0.7%  78.5%  2408609   0.7% clang::LazyGenerationalUpdatePtr::makeValue
 2192372   0.6%  79.1%  2888205   0.9% llvm::BitstreamCursor::ReadAbbrevRecord
 2097152   0.6%  79.8%  2097152   0.6% do_lookup_x
 2094845   0.6%  80.4%  2094845   0.6% _PyObject_Malloc (inline)
 2045823   0.6%  81.0%  2417955   0.7% std::vector::_M_realloc_insert
 1970252   0.6%  81.6%  1970252   0.6% clang::ImplicitCastExpr::Create
 1967305   0.6%  82.1%  1967305   0.6% clang::DeclContext::CreateStoredDeclsMap
 1942786   0.6%  82.7%  9385646   2.8% tensorflow::kernel_factory::OpKernelRegistrar::InitInternal
 1835013   0.5%  83.3%  1835013   0.5% clang::Redeclarable::DeclLink::setLatest
 1827555   0.5%  83.8%  4115915   1.2% clang::Sema::ActOnTypeName
 1767336   0.5%  84.3%  3867176   1.1% TStreamerInfo::Compile
 1761535   0.5%  84.8%  1761535   0.5% clang::ASTRecordLayout::ASTRecordLayout
 1736704   0.5%  85.4%  1736704   0.5% clang::UnresolvedLookupExpr::Create@2c2d230
 1728520   0.5%  85.9% 13959691   4.1% clang::FunctionTemplateDecl::newCommon
 1703936   0.5%  86.4%  7864320   2.3% clang::ClassTemplateSpecializationDecl::Create
 1604100   0.5%  86.8%  1604100   0.5% clang::Sema::CreateParsedType
 1572864   0.5%  87.3%  1572864   0.5% clang::TemplateArgument::CreatePackCopy
 1474654   0.4%  87.7%  1476274   0.4% clang::ASTDeclReader::AddLazySpecializations
 1407541   0.4%  88.2%  1407541   0.4% std::__cxx11::basic_string::_M_construct
 1372239   0.4%  88.6%  1372239   0.4% clang::TemplateParameterList::Create
 1355374   0.4%  89.0%  1355374   0.4% TExMap::TExMap
 1301236   0.4%  89.4%  1493304   0.4% RPCGeometryBuilderFromCondDB::build
 1286583   0.4%  89.7%  1529069   0.5% google::protobuf::internal::ArenaStringPtr::Set@3caaba
 1280356   0.4%  90.1%  1280356   0.4% THashTable::THashTable
 1228800   0.4%  90.5%  1228800   0.4% clang::NestedNameSpecifierLocBuilder::getWithLocInContext
 1167340   0.3%  90.8%  1991651   0.6% _dl_map_object_deps
 1124086   0.3%  91.2%  1124086   0.3% _dl_new_object
 1064960   0.3%  91.5%  2058562   0.6% clang::Sema::ActOnCXXNestedNameSpecifier@ff8280
```

**Comment by smuzaffar** [comment]
@gartung , is this ready to go in?

**Comment by gartung** [comment]
yes

**Comment by smuzaffar** [comment]
+externals

**Pull request merged**